upgrade_downgrade/test_dual_control_plane_upgrade_boutique.sh

```diff
@@ -0,0 +1,343 @@
+kubectl patch deployment loadgenerator -n "${TEST_NAMESPACE}" --patch '{"spec":{"template":{"metadata":{"annotations": {"sidecar.istio.io/inject": "false"}}}}}'
+kubectl rollout restart deployment loadgenerator -n "${TEST_NAMESPACE}"
+waitForPodsReady "${TEST_NAMESPACE}"
+
+# Start external traffic from fortio
+# 1. First get ingress address
+# 2. Next, use that address to fire requests at boutique shop app
+waitForIngress
+
+TRAFFIC_RUNTIME_SEC=800
+LOCAL_FORTIO_LOG=${TMP_DIR}/fortio_local.log
+EXTERNAL_FORTIO_DONE_FILE=${TMP_DIR}/fortio_done_file
+
+# TODO(su225): Move it to common as it is generic enoug
+# or probably fortio_helper.sh?
+runFortioLoadCommand() {
+  withRetries 10 10  echo_and_run fortio load -c 32 -t "${TRAFFIC_RUNTIME_SEC}"s -qps 10 -timeout 30s\
+    "http://${1}/" &> "${LOCAL_FORTIO_LOG}"
+  echo "done" >> "${EXTERNAL_FORTIO_DONE_FILE}"
+}
+
+waitForExternalRequestTraffic() {
+  echo "Waiting for external traffic to complete"
+  local attempt=0
+  while [[ ! -f "${EXTERNAL_FORTIO_DONE_FILE}" ]]; do
+    echo "attempt ${attempt}"
+    attempt=$((attempt+1))
+    sleep 10
+  done
+}
+
+# Sends external traffic from machine test is running on through external IP and ingress gateway LB.
+sendExternalRequestTraffic() {
+  writeMsg "Sending external traffic"
+  runFortioLoadCommand "${1}"
+}
+
+# Start sending traffic from outside the cluster
+sendExternalRequestTraffic "${INGRESS_ADDR}" &
+
+# Wait for some time
+# Represents stabilizing period
+STABILIZING_PERIOD=60
+writeMsg "Wait for $STABILIZING_PERIOD seconds"
+sleep "${STABILIZING_PERIOD}"
+
+# Install Target revision of Istio
+writeMsg "Deploy Istio ${TO_TAG}"
+${TO_ISTIOCTL} install -f "${TMP_DIR}/iop-control-plane.yaml" -y --revision "${TO_REVISION}"
+${TO_ISTIOCTL} install -f "${TMP_DIR}/iop-gateways.yaml" -y --revision "${TO_REVISION}"
+waitForPodsReady "${ISTIO_NAMESPACE}"
+
+# Now change labels and restart deployments one at a time
+kubectl label namespace "${TEST_NAMESPACE}" istio.io/rev-
+kubectl label namespace "${TEST_NAMESPACE}" istio.io/rev="${TO_REVISION}"
+
+# Upgrade select micro-services in the first round
+first_round=( "frontend" "redis-cart" "paymentservice" )
+for d in "${first_round[@]}"; do
+  restartDataPlane "$d" "${TEST_NAMESPACE}"
+done
+
+# **DO NOT** restart loadgenerator. We need that to run
+# continuously during upgrade so that we can see how many
+# requests fail during data plane restart after upgrade.
+#
+# TODO(su225): make a forbidden list so that we can skip restarting it
+if [[ "${TEST_SCENARIO}" == "boutique-upgrade" ]]; then
+  for d in $(kubectl get deployments -n "${TEST_NAMESPACE}" -o name | grep -v loadgenerator); do
+    deployment=$(echo "$d" | cut -d'/' -f2)
+    if [[ ! "${first_round[*]}" =~ $deployment ]]; then
+      restartDataPlane "$deployment" "${TEST_NAMESPACE}"
+    fi
+  done
+fi
+
+waitForPodsReady "${TEST_NAMESPACE}"
+
+# TODO(su225): This thing is repeating a lot. Refactor this into a function which
```

Comment:
i think you can probably leave out checking that pods are pointing to the right istiod as this sort of thing is pretty well covered in other integration tests. this will make the code a bit shorter and simpler.

Reply:
Ok. I can take out a lot of code from another test as well :) Thanks.